core/templates/variant_allocator.h

```diff
@@ -0,0 +1,294 @@
+/**************************************************************************/
+/*  variant_allocator.h                                                   */
+/**************************************************************************/
+/*                         This file is part of:                          */
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef VARIANT_ALLOCATOR_H
+#define VARIANT_ALLOCATOR_H
+
+#include "core/core_globals.h"
+#include "core/os/spin_lock.h"
+#include "core/typedefs.h"
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <atomic>
+#include <cstddef>
+#include <cstdlib>
+#include <type_traits>
+
+#if defined(__GNUC__) || defined(__clang__)
+#if (__has_builtin(__builtin_popcountll))
+#define builtin_popcountll
+#endif
+#if (__has_builtin(__builtin_ffsll))
+#define builtin_ffsll
```

Comment:
Because we don't know for certain if `uint64_t` is `unsigned long long`, I'd rather use the generic equivalents:
```suggestion
#if __has_builtin(__builtin_popcountg)
#define builtin_popcountg
#endif
#if __has_builtin(__builtin_ffsg)
#define builtin_ffsg
```

Reply:
clang doesn't have `__builtin_ffsg` or at least the version I use doesn't. I have added some static assertions to ensure the types are equivalent.

```cpp
#if defined(__GNUC__) || defined(__clang__)
#if (__has_builtin(__builtin_popcountll))
#define builtin_popcountll
static_assert(sizeof(unsigned long long) == sizeof(uint64_t), "uint64_t and unsigned long long must have the same size");
static_assert((uint64_t)(-1) == (unsigned long long)(-1), "uint64_t and unsigned long long must have the same representation");
#endif
#if (__has_builtin(__builtin_ffsll))
#define builtin_ffsll
static_assert(sizeof(long long) == sizeof(int64_t), "int64_t and long long must have the same size");
static_assert((int64_t)(-1) == (long long)(-1), "int64_t and long long must have the same representation");
#endif
#endif
```

Does this work for you?

Reply:
I like the idea, but this would entirely lockout those with `unsigned long` as their `uint64_t`. If we don't have a generic, then `__builtin_popcountll` should do the trick; I was just concerned of unnecessary conversion on those platforms (or maybe that's optimized out, I'm ignorant on compiler magic)

Alternatively, if we still want to cover all of our bases, we could try this:
```cpp
#if defined(__GNUC__) || defined(__clang__)
#if std::is_same_v<unsigned long long, uint64_t>
#if __has_builtin(__builtin_popcountll)
#define builtin_popcount_uint64 __builtin_popcountll
#endif // __has_builtin(__builtin_ffsll)
#if __has_builtin(__builtin_ffsll)
#define builtin_ffs_uint64 __builtin_ffsll
#endif // __has_builtin(__builtin_ffsll)
#else
#if __has_builtin(__builtin_popcountl)
#define builtin_popcount_uint64 __builtin_popcountl
#endif // __has_builtin(__builtin_ffsl)
#if __has_builtin(__builtin_ffsl)
#define builtin_ffs_uint64 __builtin_ffsl
#endif // __has_builtin(__builtin_ffsl)
#endif // std::is_same_v<unsigned long long, uint64_t>
#endif // defined(__GNUC__) || defined(__clang__)
```
…Though this is *really* excessive & probably not necessary.